add a CI system

samdolt opened the pull request.

samdolt commented:
You can see this github action file running here: https://github.com/samdolt/meta-codechecker/pull/1

bachp commented:
@samdolt I seems the timeout from github actions is not sufficient to run the build

samdolt commented:
> @samdolt I seems the timeout from github actions is not sufficient to run the build

@bachp : Indeed, the default timeout is only 360 minutes, but it can be set to any number between 1 minutes and 6 hours. I retried with a bigger number.

We have another problem with Github Action, it's that the cache is limited to 5GB. It will be enough for the "download" directory but not fore the SSTATES folder.

But maybe we can set up something, store the SSTATE on S3 and then using that over http.

samdolt commented:
@bachp 

So I implemted a workarround for the 360 minutes problems, by pre-generating the SSTATE on a local machine. Then I published the SSTATE online. In the CI, if the sstate folder is empty, it will automatically download the pregenerated SSTATE archive. With this scheme, the CI can build meta-codechecker in about 10 minutes.

See https://github.com/samdolt/meta-codechecker/actions/runs/1212406448

The generated report is then available as artifacts.

samdolt commented:
I added Dunfell and Gatesgarth to the CI and changed to use S3 to prefetch a locally generated SSTATE archive (needed because of Github Action limited capabilities). 

@bachp @dl9pf : Can this be merged? At least it will prevent me to break Dunfell again ;-)

bachp commented:
@samdolt I don't see how we could manage this sstate in the long run. So I rather not merge this as is.

I think you already tried http://sstate.yoctoproject.org/, what was the issue with that? Clang is not part of it?

samdolt commented:
@bachp : I tried the official sstate with the version of ubuntu provided by Github Action (without using docker) but it hasn't work. I think somehow the hash don't match the one from the sstate server.

I have looked at why, but I think it will be either:
- because somehow the distribution used change the hash, and Github Action doesn't use a "vanilla" version of ubuntu
- because adding meta-clang change the hash for all package

bachp commented:
@samdolt I looked at what others are doing:

- [meta-rust](https://github.com/meta-rust/meta-rust/blob/master/.github/workflows/main.yml): they seem to have their own runners with sstate being stored on it
- [meta-iot2050](https://github.com/siemens/meta-iot2050/blob/master/.github/workflows/main.yml) use normal runners but this probably works because it's not yocto but isar which is based on binary packages.

samdolt commented:
@bachp Having our own runner seems like the best ways to do it. We could then also persist the download and sstate dir on the machine. Maybe some low cost virtual machine like the one provided by [Scaleway](https://www.scaleway.com/en/pricing/#development-instances)
